include/hipSYCL/sycl/libkernel/atomic_fence.hpp

```diff
@@ -0,0 +1,111 @@
+ *
+ * Copyright (c) 2023 Aksel Alpay
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef HIPSYCL_ATOMIC_FENCE_HPP
+#define HIPSYCL_ATOMIC_FENCE_HPP
+
+#include "hipSYCL/sycl/libkernel/backend.hpp"
+#include "memory.hpp"
+
+namespace hipsycl {
+namespace sycl {
+namespace detail {
+
+template<memory_order, memory_scope>
+struct atomic_fence_impl
+{
+  HIPSYCL_KERNEL_TARGET
+  static void atomic_fence()
+  {}
+};
+
+template<>
+struct atomic_fence_impl<memory_order::seq_cst, memory_scope::work_group>
+{
+  HIPSYCL_KERNEL_TARGET
+  static void atomic_fence()
+  {
+    __hipsycl_if_target_hiplike(
+      __threadfence_block();
+    );
+    __hipsycl_if_target_spirv(
+    __spirv_MemoryBarrier(__spv::Scope::Workgroup,
+                          __spv::MemorySemanticsMask::SequentiallyConsistent |
+                          __spv::MemorySemanticsMask::WorkgroupMemory)
+    );
+  }
+};
+
+template<>
+struct atomic_fence_impl<memory_order::seq_cst, memory_scope::device>
+{
+  HIPSYCL_KERNEL_TARGET
+  static void atomic_fence()
+  {
+    __hipsycl_if_target_hiplike(
+      __threadfence();
+    );
+    __hipsycl_if_target_spirv(
+    __spirv_MemoryBarrier(__spv::Scope::Device,
+                          __spv::MemorySemanticsMask::SequentiallyConsistent |
+                          __spv::MemorySemanticsMask::CrossWorkgroupMemory |
+                          __spv::MemorySemanticsMask::WorkgroupMemory);
+    );
+    // TODO What about CPU?
+    // Empty __hipsycl_if_target_* breaks at compile time w/ nvc++ 22.7 or
+    // older, so comment out that statement for now.
+    //__hipsycl_if_target_host(/* todo */);
+  }
+};
```

Comment:
What is the benefit of these structs, when we do runtime if-else in `atomic_fence()` anyway?

Reply:
Little perhaps.
It was mostly to keep it similar to `mem_fence.hpp` and perhaps bring more clarity to which of the order and scope combinations we actually support.

Reply:
Completely rewritten, structs gone.